### PR TITLE
Warrior changes (latest)

### PR DESCRIPTION
**Pull request by parkscs**
I have continued working on this project since submitting the original pull request and I think it probably makes sense to close that first PR and create a new one to replace it based on the newest branch.  The warrior_changes_2 branch includes the recent work that I've done on this project.  We've had this code on the test server for some time now to collect feedback from the community and to identify bugs (of which we found and fixed a few), and at this point I think it probably makes sense to just skip warrior_changes_1 and move right to warrior_changes_2.

Because this includes the code from the original PR and I know some feedback was already given for that first PR, I've taken all the feedback from the first warrior_changes pull request and incorporated it into this branch (e.g., I removed the dead code @cizra flagged earlier today).  There's a lot here but I've done my best to keep the code separated into logical commits.

As an overview, the following PR includes the following updates to the warrior class:
- Updates to legacy warrior abilities (bodyslam, stomp, etc.) to make them worth using
- Refactoring of warrior abilities to move abilities into more logical disciplines and to spread the abilities around more evenly.  Also added 2h specialization to warriors.
- Refactoring deathstroke - this ability now deals more damage overall and scales inversely with victim health.  Added a lockout and a significant debuff to discourage tanks from using this offensive ability.
- New ability - Rally - short duration AoE buff that raises max hp/vitality
- New ability - Whirlwind - AoE attack with a lockout
- New ability - Focus Attack - a reliable additional autoattack with a lockout.
- New ability - Slam - a low damage melee attack with no lockout
- New ability - Fortify - drastically reduces incoming damage for the warrior for a very short duration.  Lockout and requires a shield.
- New ability - Bloodlust - combat and healing bonuses that stack as the warrior kills opponents.  Short duration and buffs will fall off if the warrior is out of combat for long.
- Berserk tweaks and new ability - advanced berserking - strengthens the base berserk ability and provides increased bloodlust stacks while berserking.

At this point, this is a fairly complete overhaul of the warrior class.  I anticipate there could be a few new additions in the future, but I think most changes will likely come in the form of minor buffs/nerfs/bug fixes as we get more testing/feedback.

### CODE REVIEW

**Comment by Durhamite2118** [comment]
To support this request, these changes were evaluated on the test server for ~30 days with no obvious bugs or balance issues in game play.  We are committed to the idea that if we discover an unrealized issue/bug/balance problem that we pull this back to reevaluate and correct as needed.  Thanks to Boz for his hard work.

File: code/code/cmd/cmd_rally.cc [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      tmp_victim->affectJoin(caster, &aff1, AVG_DUR_NO, AVG_EFF_YES, false);
+      tmp_victim->affectJoin(caster, &aff2, AVG_DUR_NO, AVG_EFF_YES, false);
+      tmp_victim->addToMove(modifierValue);
+      tmp_victim->addToHit(modifierValue);
+
+      found = true;
+    }
+  }
+  if (!found)
+    caster->sendTo("But, there's nobody in your group.\n\r");
```
Comment (drewhershey) [inline code/code/cmd/cmd_rally.cc:79]:
```suggestion
  bool found = false;
  for (TThing *thing : roomp->stuff) {
    auto *person = dynamic_cast<TBeing *>(thing);

    if (!person || !inGroup(*person))
      continue;

    person->sendTo("You feel revitalized and ready to continue the battle!\n\r");
    person->affectJoin(this, &aff1, AVG_DUR_NO, AVG_EFF_YES, false);
    person->affectJoin(this, &aff2, AVG_DUR_NO, AVG_EFF_YES, false);
    person->addToMove(modifierValue);
    person->addToHit(modifierValue);

    if (person != this)
      found = true;
  }

  if (!found)
    sendTo("But, there's nobody in your group.\n\r");
```

**Review by drewhershey — CHANGES_REQUESTED** [review]
Hopefully I wasn't too overbearing in my review! The raw number of comments makes it look like I found more problems than I actually did - I just tried to use lots of small, individual suggestions to make it easy to see what I was proposing be changed.

If I understand correctly how this works, for all of these suggestions except one you should be able to click either the "commit suggestion" (which will apply the suggestion as its own commit) or "add suggestion to batch" (to commit a group of suggestions together as a single commit) buttons in the individual comment to apply the suggestion straight to the code through Github. Pretty handy.

File: code/code/cmd/cmd_bodyslam.cc [hunk 2]
```diff
@@ -221,50 +221,35 @@ static int bodyslamHit(TBeing *caster, TBeing *victim)
...
-  if (caster->bSuccess(bKnown + percent, SKILL_BODYSLAM) &&
-         (i = caster->specialAttack(victim,SKILL_BODYSLAM)) && 
-         i != GUARANTEED_FAILURE &&
-         (percent < bKnown))  {
-    int modif = 1;
-    modif += (caster->getPrimaryHold() ? 0 : 2);
-    modif += (caster->getSecondaryHold() ? 0 : 1);
+  int bKnown = caster->getSkillValue(SKILL_HEADBUTT);
+  int successfulHit = caster->specialAttack(victim, SKILL_HEADBUTT);
+  int successfulSkill = caster->bSuccess(bKnown, SKILL_HEADBUTT);
```
[the remark this answers is not in the document]
Comment (parkscs) [reply code/code/cmd/cmd_bodyslam.cc:238]: That would make too much sense.  XD

**Comment by parkscs** [comment]
Pulled in your changes Drew and rebased all the commits to break it back out.  A few things I had to tweak slightly (where you referenced the variable "weapon" I think was a mistake and I changed that to reference the variable name declared above for the primary hand weapon) but otherwise I think it's pretty much as suggested.  Any further thoughts/feedback are welcome.

**Review by drewhershey — DISMISSED** [review]

**Comment by drewhershey** [comment]
Everything works well in my testing after the recent commits 👍

**Review by parkscs — COMMENTED** [review]
All looks reasonable to me.

File: code/code/cmd/cmd_deathstroke.cc [hunk 3]
```diff
@@ -5,222 +5,168 @@
...
   }
+
+  // Prevent players from attacking themselves
   if (victim == caster) {
     caster->sendTo("Do you REALLY want to kill yourself?...\n\r");
     return FALSE;
   }
+
+  // Avoid players attacking un-harmable victims
   if (caster->noHarmCheck(victim))
```
Comment (cizra) [inline code/code/cmd/cmd_deathstroke.cc:38]: No message sent? Or is that in noHarmCheck?

File: code/code/cmd/cmd_deathstroke.cc [hunk 4]
```diff
@@ -5,222 +5,168 @@
...
     if (victim->getPosition() > POSITION_DEAD) {
       act("$n's attempt at $N's vital area falls far short of hitting.", 
                   FALSE, caster, 0, victim, TO_NOTVICT);
       act("You fail to hit $N's vital area.", FALSE, caster, 0, victim, 
                   TO_CHAR);
       act("$n attempts to hit your vital area, but fails miserably.", 
                   FALSE, caster, 0, victim, TO_VICT);
-      if (caster->reconcileDamage(victim, 0,sktype) == -1)
-        return DELETE_VICT;
     }
```
Comment (cizra) [inline code/code/cmd/cmd_deathstroke.cc:117]: else what? Can it happen? Print something?

File: code/code/cmd/cmd_deathstroke.cc [hunk 5]
```diff
@@ -5,222 +5,168 @@
...
-       caster->setCharFighting(victim);
-     if (!vfight)
-       caster->setVictFighting(victim);
-  } else if (!vfight && caster->sameRoom(*victim)) {
-    // check sameRoom since may have fled from the damage
+  if (((victimSkillValue + victimSkillModifier)> 0) &&
+      victim->bSuccess(victimSkillValue + victimSkillModifier, SKILL_DEATHSTROKE) &&
+      victim->specialAttack(caster,SKILL_DEATHSTROKE) != GUARANTEED_FAILURE) {
+    // Successful counter attack
     if (victim->getPosition() > POSITION_STUNNED) {
```
Comment (cizra) [inline code/code/cmd/cmd_deathstroke.cc:142]: Kinda weird to check for skills first and stun last.. but I guess it doesn't make any difference

**Review by cizra — APPROVED** [review]

File: code/code/cmd/cmd_deathstroke.cc (hunk 4 above)
Comment (parkscs) [reply code/code/cmd/cmd_deathstroke.cc:119]: I think the else is it just continues through the function without returning.  It goes on to allow the victim a chance to counter attack down below and ultimately will return an appropriate value, but there's not specifically an else case to handle here I believe.

File: code/code/cmd/cmd_deathstroke.cc (hunk 3 above)
Comment (parkscs) [reply code/code/cmd/cmd_deathstroke.cc]: Updated to add a quick message.

File: code/code/cmd/cmd_deathstroke.cc (hunk 5 above)
Comment (parkscs) [reply code/code/cmd/cmd_deathstroke.cc:144]: Submitted an updated commit to re-order.

File: code/code/cmd/cmd_deathstroke.cc (hunk 3 above)
Comment (drewhershey) [reply code/code/cmd/cmd_deathstroke.cc]: Just FYI, noHarmCheck already sends a message based on circumstances - that's why it didn't have one here initially.

https://github.com/sneezymud/sneezymud/blob/221a9c1abea6439a7f682b220cee8cdf904633c3/code/code/misc/offense.cc#L2066-L2112

**Review by cizra — COMMENTED** [review]
This branch saw very substantial changes since my last review. I need to reread.

File: code/code/disc/disc_brawling.cc [hunk 6]
```diff
@@ -29,9 +33,13 @@ CDBrawling::CDBrawling(const CDBrawling &a) :
...
+  skHeadbutt(a.skHeadbutt),
+  skKneestrike(a.skKneestrike),
   skSpin(a.skSpin),
   skCloseQuartersFighting(a.skCloseQuartersFighting),
-  skTaunt(a.skTaunt)
+  skTaunt(a.skTaunt),
+  skTrip(a.skTrip),
+  skAdvBerserk(a.skAdvBerserk)
 {
 }
```
Comment (cizra) [inline code/code/disc/disc_brawling.cc:45]: These copy constructors and assignment operators usually don't do anything new that the compiler wouldn't generate automatically. We should get rid of them all, tho maybe not in this  PR.

Comment (drewhershey) [reply code/code/disc/disc_brawling.cc:45]: I've tested removing all those constructors from about a dozen disc_x.cc files and letting them be replaced by implicit constructors and the code compiled fine, so I think you're right that they're not needed. One big PR getting rid of them all is on my to-do list.

File: code/code/misc/limits.cc [hunk 7]
```diff
@@ -467,6 +467,22 @@ int TPerson::hitGain()
...
 
+  // Scale regen based on the number of bloodlust stacks 
+  if (affectedBySpell(SKILL_BLOODLUST)) {
+    int mod = 0;
+    affectedData *ch_affected;
+
+    for (ch_affected = affected; ch_affected; ch_affected = ch_affected->next) {
+      if (ch_affected->type == SKILL_BLOODLUST) {
+        // set the mod and remove the affect so we can add it fresh
+	mod += ch_affected->modifier;
```
Comment (cizra) [inline code/code/misc/limits.cc:478]: bad indent

File: lib/help/_skills/advanced berserking [hunk 8]
```diff
@@ -0,0 +1,5 @@
+While the average berserker represents a threat (albeit a one-dimensional one), 
+a handful of Berserker tribes have truly mastered the art.  These advanced berserkers
+maintain all the reckless abandon of a typical berserker but with even further
+heightened combat prowess.  Moreover, particularly skilled berserkers can work themselves
+into a bloodlust the longer they fight.
```
Comment (cizra) [inline lib/help/_skills/advanced berserking:5]: There's some grammar missing here. They can work themselves into a _more intense_ bloodlust the longer they fight, something like that?

File: lib/help/_skills/bloodlust [hunk 9]
```diff
@@ -0,0 +1,17 @@
+Bloodlust represents a warrior's love of battle and their desire to
+continue the fight.  With each foe vanquished, the warrior's bloodlust
+continues to grow, increasing their effectiveness in battle.  Moreover,
+the accompanying adrenaline rush can allow the warrior to continue to 
+fight and sustain themselves, well beyond the stamina and vitality of
+the average adventurer.  However, this fervor of battle can fade as quickly
+as it comes on once a battle has ended.
+
+Rumor has it that some tribes of northern Berskerers have further mastered
```
Comment (cizra) [inline lib/help/_skills/bloodlust:9]: Oooh, opportunity to tie this into racial subcategories!

File: lib/help/_skills/whirlwind [hunk 10]
```diff
@@ -0,0 +1,9 @@
+Through their mastery of armed combat, warriors are able to swing their
+weapons in wide arcs, striking multiple opponents simultaneously.  Such
+whirlwind attacks can only be attempted periodically, as they can throw
+the warrior off-balance and are draining on even the most veteran warriors.
+
+
+Technical description: Whirlwind attempts to strike all non-grouped entities
+within the room.  A skill-success check is applied for attempting the maneuver,
+and then a separate hit-check is applied for each and every non-grouped entity.
```
Comment (cizra) [inline lib/help/_skills/whirlwind:9]: Changes of this caliber merit an update of the news file

**Review by cizra — CHANGES_REQUESTED** [review]

File: code/code/disc/disc_brawling.cc (hunk 6 above)
Comment (parkscs) [reply code/code/disc/disc_brawling.cc:45]: Happy to help if I can with removing them/testing, but would prefer to do so in a new PR.

File: lib/help/_skills/advanced berserking (hunk 8 above)
Comment (parkscs) [reply lib/help/_skills/advanced berserking:5]: The grammar is correct, but I can reword to make it clearer.  The intent is to convey that bloodlust stacks can be gained in two ways - killing an enemy or, with advanced berserking, over time while fighting.

File: lib/help/_skills/bloodlust (hunk 9 above)
Comment (parkscs) [reply lib/help/_skills/bloodlust:9]: We were floating around some ideas for synergies with various races and shapeshifts recently - definitely something to consider in the future, especially if it's not game-changing buffs/advantages but just adds a bit of flavor.

File: lib/help/_skills/whirlwind (hunk 10 above)
Comment (parkscs) [reply lib/help/_skills/whirlwind:9]: Pushed an updated news file - will need to update the date in the entry before this gets final approval/merged.

File: code/code/misc/limits.cc (hunk 7 above)
Comment (parkscs) [reply code/code/misc/limits.cc:478]: I think I addressed the concern?  I had some tab characters in a couple indents, replaced with all space characters.  Let me know if the issue was something else.

Comment (cizra) [reply code/code/misc/limits.cc]: Yep, looks fine now.

**Review by cizra — APPROVED** [review]

**Pull request merged**